docs/DesignDocs/CustomAttestation_V3.md

```diff
@@ -0,0 +1,423 @@
+Custom Attestation Data Formats for Open Enclave V3 Updates
+================
+
+As custom attestation plugins are developed and features are requested, new APIs were added to:
+
+1. Query built-in SGX attesters and verifiers to be registered.
+2. Select a format of the evidence that the registered attesters can generate.
+3. Query evidence parameters to be used as optional inputs when generating evidences.
+
+The following API was affected:
+
+1. Built-in SGX attesters and verifiers
+
+This document describes the added and affected APIs as well as keeping the existing support for version 2 of the API. This document assumes the user is familiar with the SGX attestation and [custom attestation](https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/CustomAttestation.md) with plugins.
+
+Motivation
+----------
+
+New APIs were added to help customers who would like to extend Open Enclave's current attestation structure with extra information while keep TEE-specific information within attester and verifier plugins.
+
+1. The application can query built-in SGX attesters and verifiers to be registered. This allows the plugin library to return attesters and verifiers based on configuration.
+
+```C
+   // Old
+   oe_register_attester(oe_sgx_attesterX(), NULL, 0);
+   oe_register_attester(oe_sgx_attesterY(), NULL, 0);
+   oe_register_attester(oe_sgx_attesterZ(), NULL, 0);
+
+   // New
+   oe_sgx_get_attesters(&attesters, &attesters_length);
```

Comment:
(Same applies to both attester and verifier below)
1. oe_sgx_get_verifiers() returns a set of verifiers that may not be known at build-time. For example, the same code could run on VM with different configuration or back-end setup.
2. We want to preserve the original design of oe_register_verifier() that takes configuration (2nd parameter) as input.
3. oe_sgx_get_verifiers() is a replacement of existing API oe_sgx_plugin_verifier() that is to be deprecated. Eventually SGX attesters/verifiers should be in its own library but until then, the API naming remains oe_sgx_*.

---
In reply to: [393414859](https://github.com/openenclave/openenclave/pull/2621#discussion_r393414859) [](ancestors = 393414859)

Reply:
> 1. oe_sgx_get_verifiers() returns a set of verifiers that may not be known at build-time. For example, the same code could run on VM with different configuration or back-end setup.

How? It's not described in this doc. This is before registration of the verifiers just to get the UUID for registration, so my expectation is the only way that works is a hard coded list of values returned by the non-extensible `oe_sgx_get_verifiers()` (That's how the individual `oe_sgx_verifier*()` functions it replaces were implemented). 

> 2. We want to preserve the original design of oe_register_verifier() that takes configuration (2nd parameter) as input.

This is a problem because it means that a dev will still usually have to know exactly which format IDs need to be configured in special ways. What are the benefits over this strawman alternative?

* Dev calls `oe_register_verifier_plugins()` which attempts to register all plugins linked into the app, ignores unsupported ones, and reports the IDs of all plugins that require configuration.
* Dev invokes `oe_configure_verifier_plugin()` for plugins that have required (or even optional additional) configuration.

> 3. oe_sgx_get_verifiers() is a replacement of existing API oe_sgx_plugin_verifier() that is to be deprecated. Eventually SGX attesters/verifiers should be in its own library but until then, the API naming remains oe_sgx_*.

What time line is "eventually" on? This is a public API surface and I would like to avoid churning it as much as possible. If they will be deprecated, and it doesn't seem like the proposed replacement is significantly better, then why replace this now instead of leaving it until it's entirely removed?